libjcat/jcat-libcrypto-pkcs7-engine.c

```diff
@@ -0,0 +1,484 @@
+			}
+
+			issuer_size = BIO_get_mem_data(issuer_bio, &issuer_string);
+			if (issuer_size > 0 && issuer_string != NULL) {
+				g_string_overwrite_len(authority_newest,
+						       0,
+						       issuer_string,
+						       issuer_size);
+			}
+		}
+	}
+
+	/* success */
+	return JCAT_RESULT(g_object_new(JCAT_TYPE_RESULT,
+					"engine",
+					engine,
+					"timestamp",
+					timestamp_newest,
+					"authority",
+					authority_newest->str,
+					NULL));
+}
+
+static JcatResult *
+jcat_libcrypto_pkcs7_engine_self_verify(JcatEngine *engine,
+					GBytes *blob,
+					GBytes *blob_signature,
+					JcatVerifyFlags flags,
+					GError **error)
+{
+	g_autofree gchar *filename = NULL;
+	g_autoptr(GBytes) cert_blob = NULL;
+	g_autoptr(X509) crt = NULL;
+
+	filename =
+	    g_build_filename(jcat_engine_get_keyring_path(engine), "pki", "client.pem", NULL);
+	cert_blob = jcat_get_contents_bytes(filename, error);
+	if (cert_blob == NULL)
+		return NULL;
+	crt = jcat_libcrypto_pkcs7_load_crt_from_blob_pem(cert_blob, error);
+	if (crt == NULL)
+		return NULL;
+
+	return jcat_libcrypto_pkcs7_engine_verify(engine, blob, blob_signature, crt, flags, error);
+}
+
+static JcatResult *
+jcat_libcrypto_pkcs7_engine_pubkey_verify(JcatEngine *engine,
+					  GBytes *blob,
+					  GBytes *blob_signature,
+					  JcatVerifyFlags flags,
+					  GError **error)
+{
+	return jcat_libcrypto_pkcs7_engine_verify(engine, blob, blob_signature, NULL, flags, error);
+}
+
+static JcatBlob *
+jcat_libcrypto_pkcs7_engine_pubkey_sign(JcatEngine *engine,
+					GBytes *blob,
+					GBytes *cert,
+					GBytes *privkey,
+					JcatSignFlags flags,
+					GError **error)
+{
+	g_autoptr(EVP_PKEY) key = NULL;
+	g_autoptr(X509) crt = NULL;
+	g_autoptr(CMS_ContentInfo) cms_ci = NULL;
+	g_autoptr(BIO) blob_bio = NULL;
+	g_autoptr(BIO) sig_bio = NULL;
+	g_autoptr(GString) rewrap = NULL;
+	gchar *bio_buf;
+	gsize bio_len;
+	guint signing_flags = CMS_BINARY | CMS_DETACHED | CMS_NOSMIMECAP;
+
+	/* nothing to do */
+	if (g_bytes_get_size(blob) == 0) {
+		g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT, "nothing to do");
+		return NULL;
+	}
+
+	/* load keys */
+	key = jcat_libcrypto_pkcs7_load_privkey_from_blob_pem(privkey, error);
+	if (key == NULL)
+		return NULL;
+	crt = jcat_libcrypto_pkcs7_load_crt_from_blob_pem(cert, error);
+	if (crt == NULL)
+		return NULL;
+
+	/* get the digest algorithm from the public key */
+	// TODO: The gnutls uses the keys "preferred_hash_algorithm". Port that if necissary.
+	// gchar mdname[256] = {0};
+	// gint default_digest_nid = 0;
+	// EVP_PKEY_get_default_digest_name(key, mdname, sizeof(mdname));
+	// EVP_PKEY_get_default_digest_nid(key, &default_digest_nid);
+	// g_debug("pubkey_sign pref digest %s %d", mdname, default_digest_nid);
+
+	/* sign data */
+	// TODO: Implement these sign ing flags
+	// TODO: CMS_NO_SIGNING_TIME was added in OpenSSL 3.5
+	// if (!(flags & JCAT_SIGN_FLAG_ADD_TIMESTAMP))
+	//	signing_flags |= CMS_NO_SIGNING_TIME;
```

Comment:
can we add some meson +config.h `HAVE_LIBCRYPTO_NO_SIGNING_TIMESTAMP` for that?

Reply:
I added `#if OPENSSL_VERSION_PREREQ(3, 5)`. Would testing it in meson be better?
https://github.com/hughsie/libjcat/pull/177/commits/b58a1d29455fe46a6d431805db9c6dc28f2ecc6a